File: 547-number-of-provinces/547-number-of-provinces.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        N = len(isConnected)
        parent = [i for i in range(N)]
        rank, res = [1] * N, N
        
        def find(nd):
            while nd != parent[nd]:
                nd = parent[nd]
            return nd
        
        def union(n1, n2):
            p1, p2 = find(n1), find(n2)
            
            if p1 == p2: return 0
            if rank[p1] < rank[p2]:
                parent[p1] = p2
                rank[p2] += rank[p1]
            else:
                parent[p2] = p1
                rank[p1] += rank[p2]
            return 1
        
        for i in range(N):
            for j in range(N):
                if isConnected[i][j]:
                    res -= union(i,j)
                    
        return res

File: 547-number-of-provinces/547-number-of-provinces.cpp
class Solution {
public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n = isConnected.size(), cnt = n;
        vector<int>link(n), size(n, 1);
        iota(link.begin(), link.end(), 0);
        
        function<int(int)> find = [&](int x) {
            while (x != link[x])
                x = link[x];
            return x;
        };
        
        function<void(int, int)> unite = [&](int a, int b) {
            int roota = find(a), rootb = find(b);
            if (roota == rootb)
                return;
            if (size[a] < size[b])
                swap(a,b);
            link[rootb] = roota;
            size[roota] += size[rootb];
            cnt--;
        };
        
        for (int i = 0; i < n; i++)
            for (int j = i + 1; j < n; j++)
                if (isConnected[i][j])
                    unite(i, j);
                
        return cnt;
    }
};